File: myExpenses/src/test/java/org/totschnig/myexpenses/export/qif/QifUtilTwoSidesOfTheSameTransferTest.java
package org.totschnig.myexpenses.export.qif;

import junit.framework.Assert;
import junit.framework.TestCase;

import java.math.BigDecimal;
import java.util.Date;

public class QifUtilTwoSidesOfTheSameTransferTest extends TestCase {

  public void testShouldMatchTwoSidesOfSameTransfer() {
    Date now = new Date(System.currentTimeMillis());

    QifAccount fromAccount = new QifAccount();
    fromAccount.memo = "Konto 1";
    QifAccount toAccount = new QifAccount();
    toAccount.memo = "Konto 2";
    QifTransaction fromTransaction = new QifTransaction();
    fromTransaction.toAccount = toAccount.memo;
    fromTransaction.date = now;
    fromTransaction.amount = new BigDecimal(5);
    QifTransaction toTransaction = new QifTransaction();
    toTransaction.toAccount = fromAccount.memo;
    toTransaction.date = now;
    toTransaction.amount = new BigDecimal(-5);
    Assert.assertTrue(QifUtils.twoSidesOfTheSameTransfer(fromAccount, fromTransaction, toAccount, toTransaction));
  }

  public void testShouldDistinguishNonMatchingAmounts() {
    Date now = new Date(System.currentTimeMillis());

    QifAccount fromAccount = new QifAccount();
    fromAccount.memo = "Konto 1";
    QifAccount toAccount = new QifAccount();
    toAccount.memo = "Konto 2";
    QifTransaction fromTransaction = new QifTransaction();
    fromTransaction.toAccount = toAccount.memo;
    fromTransaction.date = now;
    fromTransaction.amount = new BigDecimal(5);
    QifTransaction toTransaction = new QifTransaction();
    toTransaction.toAccount = fromAccount.memo;
    toTransaction.date = now;
    toTransaction.amount = new BigDecimal(-6);
    Assert.assertFalse(QifUtils.twoSidesOfTheSameTransfer(fromAccount, fromTransaction, toAccount, toTransaction));
  }

  public void testShouldDistinguishNonMatchingDates() {
    Date now = new Date(System.currentTimeMillis());

    QifAccount fromAccount = new QifAccount();
    fromAccount.memo = "Konto 1";
    QifAccount toAccount = new QifAccount();
    toAccount.memo = "Konto 2";
    QifTransaction fromTransaction = new QifTransaction();
    fromTransaction.toAccount = toAccount.memo;
    fromTransaction.date = now;
    fromTransaction.amount = new BigDecimal(5);
    QifTransaction toTransaction = new QifTransaction();
    toTransaction.toAccount = fromAccount.memo;
    toTransaction.date = new Date(System.currentTimeMillis()-100000);
    toTransaction.amount = new BigDecimal(-5);
    Assert.assertFalse(QifUtils.twoSidesOfTheSameTransfer(fromAccount, fromTransaction, toAccount, toTransaction));
  }

  public void testShouldDistinguishNonMatchingAccounts() {
    Date now = new Date(System.currentTimeMillis());

    QifAccount fromAccount = new QifAccount();
    fromAccount.memo = "Konto 1";
    QifAccount toAccount = new QifAccount();
    toAccount.memo = "Konto 2";
    QifTransaction fromTransaction = new QifTransaction();
    fromTransaction.toAccount = toAccount.memo;
    fromTransaction.date = now;
    fromTransaction.amount = new BigDecimal(5);
    QifTransaction toTransaction = new QifTransaction();
    toTransaction.toAccount = "Konto 3";
    toTransaction.date = now;
    toTransaction.amount = new BigDecimal(-5);
    Assert.assertFalse(QifUtils.twoSidesOfTheSameTransfer(fromAccount, fromTransaction, toAccount, toTransaction));
  }
}

File: myExpenses/src/test/java/org/totschnig/myexpenses/export/qif/QifUtilTwoSidesOfTheSameTransferTest.kt
package org.totschnig.myexpenses.export.qif

import com.google.common.truth.Truth
import junit.framework.TestCase
import org.totschnig.myexpenses.export.qif.QifUtils.twoSidesOfTheSameTransfer
import org.totschnig.myexpenses.io.ImportAccount
import org.totschnig.myexpenses.io.ImportTransaction
import java.math.BigDecimal
import java.util.Date

class QifUtilTwoSidesOfTheSameTransferTest : TestCase() {


    fun testShouldMatchTwoSidesOfSameTransfer() {
        val now = Date(System.currentTimeMillis())
        val fromAccount = ImportAccount(memo = "Konto 1")
        val toAccount = ImportAccount(memo = "Konto 2")
        val fromTransaction = ImportTransaction.Builder()
            .toAccount(toAccount.memo)
            .date(now)
            .amount(BigDecimal(5))
            .build()!!
        val toTransaction = ImportTransaction.Builder()
            .toAccount(fromAccount.memo)
            .date(now)
            .amount(BigDecimal(-5))
            .build()!!
        Truth.assertThat(twoSidesOfTheSameTransfer(
            fromAccount,
            fromTransaction,
            toAccount,
            toTransaction
        )).isTrue()
    }

    fun testShouldDistinguishNonMatchingDates() {
        val now = Date(System.currentTimeMillis())
        val fromAccount = ImportAccount(memo = "Konto 1")
        val toAccount = ImportAccount(memo = "Konto 2")
        val fromTransaction = ImportTransaction.Builder()
            .toAccount(toAccount.memo)
            .date(now)
            .amount(BigDecimal(5))
            .build()!!
        val toTransaction = ImportTransaction.Builder()
            .toAccount(fromAccount.memo)
            .date(Date(System.currentTimeMillis() - 100000))
            .amount(BigDecimal(-5))
            .build()!!
        Truth.assertThat(twoSidesOfTheSameTransfer(
            fromAccount,
            fromTransaction,
            toAccount,
            toTransaction
        )).isFalse()
    }

    fun testShouldDistinguishNonMatchingAccounts() {
        val now = Date(System.currentTimeMillis())
        val fromAccount = ImportAccount(memo = "Konto 1")
        val toAccount = ImportAccount(memo = "Konto 2")
        val fromTransaction = ImportTransaction.Builder()
            .toAccount(toAccount.memo)
            .date(now)
            .amount(BigDecimal(5))
            .build()!!
        val toTransaction = ImportTransaction.Builder()
            .toAccount("Konto 3")
            .date(now)
            .amount(BigDecimal(-5))
            .build()!!
        Truth.assertThat(twoSidesOfTheSameTransfer(
            fromAccount,
            fromTransaction,
            toAccount,
            toTransaction
        )).isFalse()
    }
}